core/src/com/unciv/ui/screens/mapeditorscreen/MapEditorWesnothImporter.kt

```diff
@@ -0,0 +1,156 @@
+                    ToastPopup("Could not load map!", editorScreen)
+                }
+            }
+        }
+    }
+
+    private fun String.parse(): TileMap {
+        // first we need to know the size. Wesnoth maps have a non-playable border - exclude.
+        val lines = lineSequence().filter { it.isNotBlank() }.toList()
+        val height = lines.size - 1
+        val width = if (height <= 0) 0 else lines[0].split(',').size - 2
+        if (width <= 0) throw UncivShowableException("That map is invalid!")
+
+        val map = TileMap(width, height, ruleset, false)
+        map.mapParameters.apply {
+            type = MapType.empty
+            shape = MapShape.rectangular
+            mapSize = MapSizeNew(width, height)
+        }
+
+        val colOffset = 1 + width / 2
+        val rowOffset = height / 2
+        for ((row, line) in lines.withIndex()) {
+            for ((column, cellCode) in line.split(',').withIndex()) {
+                val effectiveRow = rowOffset - row + column % 2
+                val pos = HexMath.getTileCoordsFromColumnRow(column - colOffset, effectiveRow)
+                if (!map.contains(pos)) continue
+                map[pos].paintFromWesnothCode(cellCode.trim(), map)
+            }
+        }
+
+        return map
+    }
+
+    private fun Tile.paintFromWesnothCode(cellCode: String, map: TileMap) {
+        // See https://wiki.wesnoth.org/TerrainCodesWML
+        val matches = parseTerrain.matchEntire(cellCode)
+            ?: throw UncivShowableException("{That map is invalid!}\n{(\"[$cellCode]\" does not conform to TerrainCodesWML)}")
+        val start = matches.groups["start"]?.value
+        val base = matches.groups["base"]!!.value  // This capture is not optional in the pattern
+        val layer = matches.groups["layer"]?.value
+        val allStrings = translateTerrainWML(base) + translateTerrainWML(layer) + Constants.grassland
+        val allObjects = allStrings
+            .sortedBy { it != Constants.hill }
+            .mapNotNull { ruleset.tileImprovements[it] ?: ruleset.terrains[it] }
+            .toList()
+        baseTerrain = allObjects.first { it is Terrain && it.type.isBaseTerrain }.name
+        val features = allObjects.filter { it is Terrain && it.type == TerrainType.TerrainFeature }.map { it.name }.distinct()
+        if (features.isNotEmpty()) {
+            setTerrainTransients() // or else can't setTerrainFeatures as baseTerrainObject may be uninitialized
+            setTerrainFeatures(features)
+        }
+        allObjects.firstOrNull { it is TileImprovement }?.apply { improvement = name }
+        if (start == null) return
+        map.addStartingLocation(Constants.spectator, this)
+    }
+
+    private fun translateTerrainWML(code: String?) = sequence {
+        if (code == null) return@sequence
+        translationCodes[code]?.also { yieldAll(it) }
+        if (code.length >= 3) // kotlin slice is unsafe, unlike python slice
```

Comment:
I refuse to believe there isn't a better way to do this. What are you trying to accomplish?